### PR TITLE
[AMDGPU][GlobalISel] Fix load/store of pointer vectors, buffer.*.pN

### PR DESCRIPTION
**Pull request by krzysz00**
Certain pointer address spaces were not being correctly handled by the GlobalISel lowering for buffer_load and buffer_store.

1. ptr addrspace(1) and addrspace(4) did not have rewrite patterns defined for them, while p0 did, since those pointer types weren't in the list of types that was iterated to form the patterns.
2. Vectors of pointers need to be bitcast to vectors of the corresponding scalars, since there doesn't seem to be a good way to define the rewrite patterns for buffer_load/store of those types

The need to bitcast vectors of pointers was also revealed to affect ordinary `G_LOAD` and `G_STORE` in some cases, so `shouldBitcastLoadStore()` has been fixed  to handle it properly.

### CODE REVIEW

**Comment by arsenm** [comment]
> 2. Vectors of pointers need to be bitcast to vectors of the corresponding scalars, since there doesn't seem to be a good way to define the rewrite patterns for buffer_load/store of those types

I'd hope you can declare them with VTVec and PtrValueType, but it probably requires new GlobalISelEmitter support. VTVec wants the enum value from MVT for some reason

File: llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp [hunk 1]
```diff
@@ -5805,6 +5806,10 @@ Register AMDGPULegalizerInfo::fixStoreSourceType(
   if (hasBufferRsrcWorkaround(Ty))
     return castBufferRsrcToV4I32(VData, B);
 
+  if (shouldBitcastLoadStoreType(ST, Ty, MemTy) || Ty.isPointerVector()) {
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:5809]: I think the Ty.isPointerVector() is redundant, these should be handled under the loadStoreBitcastWorkaround case

File: llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp [hunk 2]
```diff
@@ -5982,9 +5994,21 @@ bool AMDGPULegalizerInfo::legalizeBufferLoad(MachineInstr &MI,
   // Make addrspace 8 pointers loads into 4xs32 loads here, so the rest of the
   // logic doesn't have to handle that case.
   if (hasBufferRsrcWorkaround(Ty)) {
+    Observer.changingInstr(MI);
     Ty = castBufferRsrcFromV4I32(MI, B, MRI, 0);
+    Observer.changedInstr(MI);
     Dst = MI.getOperand(0).getReg();
+    B.setInsertPt(B.getMBB(), MI);
   }
+  if (shouldBitcastLoadStoreType(ST, Ty, MemTy) || Ty.isPointerVector()) {
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:6003]: Ty.isPointerVector should be redundant here too

File: llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp (hunk 1 above)
Comment (krzysz00) [reply llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:5809]: It isn't. All the pointers vectors fall to the `Size <= 64` case and so don't get bitcast

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:5809]: This sounds like just a bug, and this should be folded into shouldBitcastLoadStoreType. It already tries to handle this case, just too late

Comment (krzysz00) [reply llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:5809]: Except that a regular G_LOAD or G_STORE can handle vectors of pointers just fine, and moving the condition changes a bunch of tests

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:5809]: I don't think they do work fine. The buffer cases should behave identically to G_LOAD and G_STORE, and both currently rely on the bitcast legalization.

Based on the 64-bit check, the 128-bit <2 x ptr> cases work. The 64-bit cases, e.g. <2 x ptr addrspace(3)>  fail:
https://godbolt.org/z/4vWacff1d

So yes, tests should change and start working

**Review by arsenm — APPROVED** [review]
This does show we have missing end to end test coverage for vectors of pointers

**Comment by chapuni** [comment]
Excuse me, I've reverted this. For example, https://lab.llvm.org/buildbot/#/builders/63/builds/1865

**Comment by ppenzin** [comment]
Local failures also with the patch reapplied, as well as that builder is still failing. It is quite odd, not sure what is causing that.

**Comment by chapuni** [comment]
Reverted again.

I've investigated further. Tests will fail with `-DNDEBUG`.

`reportGISelFailure()` is called only in +Asserts.
https://github.com/llvm/llvm-project/commit/2b94972eb981090e0289ba56edd22b311e9fd6f7#diff-764f8c5ac9f21f864c5f677d103b945de05af957c7f993eebeb0bce8fbd25125R88
It marks a property.
https://github.com/llvm/llvm-project/commit/7c6ca18fffdcf935b3b38ffc8697180d60511ffe#diff-7b1e350b8a96379b42928d6d373d502796b98500bbfe87d3b7abbcc08e1b8bc5R228

A quick fix will be mark tests with `REQUIRES: asserts`.
I didn't notice why the result was affected by assertions.

**Pull request merged**